packages/firmware_notifier/lib/firmware_notifier.dart

```diff
@@ -0,0 +1,37 @@
+import 'package:collection/collection.dart';
+import 'package:desktop_notifications/desktop_notifications.dart';
+import 'package:fwupd/fwupd.dart';
+
+Future<Map<FwupdDevice, FwupdRelease>> getUpdates() async {
+  final fwupdClient = FwupdClient();
+  await fwupdClient.connect();
+
+  final devices = await fwupdClient.getDevices();
+  final updates = {
+    for (final device in devices)
+      device: (await fwupdClient.getReleases(device.deviceId).catchError(
+              (_) => <FwupdRelease>[],
+              test: (e) =>
+                  e is FwupdNothingToDoException ||
+                  e is FwupdNotSupportedException))
+          .firstWhereOrNull(
+        (r) => r.flags.contains(FwupdReleaseFlag.isUpgrade),
+      )
+  }..removeWhere((_, update) => update == null);
+  await fwupdClient.close();
+  return Map<FwupdDevice, FwupdRelease>.from(updates);
+}
+
+Future<List<Notification>> sendUpdateNotifications(
+    NotificationsClient client) async {
+  final updates = await getUpdates();
```

Comment:
nit-pick: the code would become testable if you'd move this to main() and pass the list of updates as an argument :wink:

Reply:
:+1: